### PR TITLE
8253525: Implement getInstanceSize/sizeOf intrinsics

### PR DESCRIPTION
**Pull request by shipilev**
This is fork off the SizeOf JEP, JDK-8249196. There is already the entry point in JDK that can use the intrinsic like this: `Instrumentation.getInstanceSize`. Therefore, we can implement the C1/C2 intrinsic now, hook it up to `Instrumentation`, and let the tools use that fast path today.

With this patch, JOL is able to be close to `deepSizeOf` implementation from SizeOf JEP. 

Example performance improvements for sizing up a custom linked list:

```
Benchmark                     (size)  Mode  Cnt       Score      Error  Units

# Default
LinkedChainBench.linkedChain       1  avgt    5     705.835 ±    8.051  ns/op
LinkedChainBench.linkedChain      10  avgt    5    3148.874 ±   37.856  ns/op
LinkedChainBench.linkedChain     100  avgt    5   28693.256 ±  142.254  ns/op
LinkedChainBench.linkedChain    1000  avgt    5  290161.590 ± 4594.631  ns/op

# Instrumentation attached, no intrinsics
LinkedChainBench.linkedChain       1  avgt    5    159.659 ±   19.238  ns/op
LinkedChainBench.linkedChain      10  avgt    5    717.659 ±   22.540  ns/op
LinkedChainBench.linkedChain     100  avgt    5   7739.394 ±  111.683  ns/op
LinkedChainBench.linkedChain    1000  avgt    5  80724.238 ± 2887.794  ns/op

# Instrumentation attached, new intrinsics
LinkedChainBench.linkedChain       1  avgt    5     95.254 ±   0.808  ns/op
LinkedChainBench.linkedChain      10  avgt    5    261.564 ±   8.524  ns/op
LinkedChainBench.linkedChain     100  avgt    5   3367.192 ±  21.128  ns/op
LinkedChainBench.linkedChain    1000  avgt    5  34148.851 ± 373.080  ns/op
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8253525](https://bugs.openjdk.java.net/browse/JDK-8253525): Implement getInstanceSize/sizeOf intrinsics

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 482c2f245ec2acf1d606a01002212da35272fd9d
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to 1b7290a32a2df6eff7efafbbef84d5d864f51d7e

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/650/head:pull/650`
`$ git checkout pull/650`

### CODE REVIEW

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Always run graalunit testing with new intrinsics.
You need to adjust Graal test:
src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java

**Comment by shipilev** [comment]
I just ran `CONF=linux-x86_64-server-fastdebug make clean run-test TEST=compiler/graalunit` without problems even without the `CheckGraalIntrinsics.java` changes. Does the test actually work?

**Comment by shipilev** [comment]
@vnkozlov I added the new block in `CheckGraalIntrinsics.java`, and made sure it passes the entire `compiler/graalunit` with [JDK-8254785](https://bugs.openjdk.java.net/browse/JDK-8254785) applied.

File: test/jdk/java/lang/instrument/GetObjectSizeTest.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2003, 2015, Oracle and/or its affiliates. All rights reserved.
```
Comment (vnkozlov) [inline test/jdk/java/lang/instrument/GetObjectSizeTest.java:2]: You can't replace Copyright lines of one company with other.

File: src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java [hunk 2]
```diff
@@ -524,6 +524,11 @@ public CheckGraalIntrinsics() {
                                 "jdk/internal/vm/vector/VectorSupport.unaryOp(ILjava/lang/Class;Ljava/lang/Class;ILjava/lang/Object;Ljava/util/function/Function;)Ljava/lang/Object;");
         }
 
+        if (isJDK16OrHigher()) {
+            add(toBeInvestigated,
+                                "sun/instrument/InstrumentationImpl.getObjectSize0(JLjava/lang/Object;)J");
+        }
+
```
Comment (vnkozlov) [inline src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java:531]: Agree with this.

**Review by vnkozlov — CHANGES_REQUESTED** [review]

File: test/jdk/java/lang/instrument/GetObjectSizeTest.java (hunk 1 above)
Comment (shipilev) [reply test/jdk/java/lang/instrument/GetObjectSizeTest.java:2]: Well, I am replacing the entire file. There is a recent precedent of the similar change [here](https://github.com/openjdk/jdk/commit/6d13c766#diff-0daf75421f8fdb55a5640742ef6f12730fe1b370cc864311c188ad99996b51fe). Either that should be reversed, or this one accepted.

Comment (shipilev) [reply test/jdk/java/lang/instrument/GetObjectSizeTest.java:2]: ...or I can put new test into a separate file. But the existing test is quite inferior compared to the new one.

Comment (vnkozlov) [reply test/jdk/java/lang/instrument/GetObjectSizeTest.java:2]: It was mistake in 8253191 (I file bug). If you modify existing file (even if you keep only test name the same) you have to preserve original Copyright and add new Copyright line. You don't need create new file.
We have a lot of cases with 2 or more Copyright lines - it is normal:
https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/vectorization/TestVectorsNotSavedAtSafepoint.java

Comment (vnkozlov) [reply test/jdk/java/lang/instrument/GetObjectSizeTest.java:2]: I file 8255067 to restore Copyright line in TestUnsignedByteCompare.java test file.

Comment (shipilev) [reply test/jdk/java/lang/instrument/GetObjectSizeTest.java]: I made sure the new test is in the new file.

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by shipilev** [comment]
Thanks for review, @kvn! I would also like a review from someone from serviceability.

/reviewers 2

**Comment by shipilev** [comment]
Friendly reminder.

**Comment by sspitsyn** [comment]
Hi Aleksey,

I've not looked at the compiler generated code.
The fix looks okay to me.
I have a question and a couple of minor suggestions on new test.
Q: Why the value of ITERS is that big? What is the need to have this number of iterations?
Also, I do not like one-letter identifiers, especially if they are not local.
Could you, please, replace identifiers R and A with some short versions that give a hint.
Something like REFSIZE and ALIGNMENT would be good enough.
Also, what tests did you run to make sure no regression is introduced?

Thanks,
Serguei

**Comment by shipilev** [comment]
> I have a question and a couple of minor suggestions on new test.
> Q: Why the value of ITERS is that big? What is the need to have this number of iterations?

The test verifies the answer does not change if we hit JIT compilers in that code. Since we are doing C1/C2 intrinsics, we need to execute the loops more than compilation-threshold times. Since the current threshold is about 100K, doing 1M iterations seems good: it is well past the compilation threshold times, and there is time to re-enter the newly compiled method. The test run time is still sane, 1 minute on my Linux x86_64 fastdebug. I can do 200K iterations and -Xbatch instead, though, see new change. This drops the test run time to 30 seconds.

> Also, I do not like one-letter identifiers, especially if they are not local.
> Could you, please, replace identifiers R and A with some short versions that give a hint.
> Something like REFSIZE and ALIGNMENT would be good enough.

Renamed to `REF_SIZE` and `OBJ_ALIGN` instead.

> Also, what tests did you run to make sure no regression is introduced?

Old code calls into `oop::size()` to get the object size. That method decodes the object's layout helper. So when we replace it with intrinsic, we now have to test the different shapes of the layout helper and varying conditions for that decoding. So the new test tries to cover the comprehensive matrix:
 - the usual object shapes: objects, primitive arrays, object arrays;
 - different compressed oops modes that affect reference sizes;
 - different object alignment modes that affect object sizes;
 - different compilation modes: interpreter, C1, C2;
 - special paths like carrying special bits in layout helper for allocation slow-paths;

I know that test is sensitive to compiler intrinsics bugs, as I used these tests to develop the intrinsics. If you inject simple off-by-one bugs into current C1/C2 intrinsics, new test catches that.

The additional safety comes from the somewhat loose API requirement: it is specified to return the guess, and that guess might as well be wrong (not overly wrong though, for a quality implementation).

**Review by sspitsyn — APPROVED** [review]
Aleksey,

Thank you for the update!
It looks good to me.
One more nit, I forgot to list in my previous comment, is uneeded '()' around comparisons:
`+    static final int REF_SIZE = ((compressedOops == null) || (compressedOops == true)) ?  4 : 8;`

Thanks,
Serguei

**Comment by shipilev** [comment]
> One more nit, I forgot to list in my previous comment, is uneeded '()' around comparisons:
> `+ static final int REF_SIZE = ((compressedOops == null) || (compressedOops == true)) ? 4 : 8;`

Right. Fixed that. Thanks!

**Comment by shipilev** [comment]
Thanks @vnkozlov and @sspitsyn!

/integrate

**Pull request closed without merging**